File: term_4 (F#, FP)/homework_5/homework_5/Stack.fs
namespace Homework_5

module Stack =

    /// Стек
    type Stack<'a> =
        | Empty
        | Stack of 'a * Stack<'a>
    
    /// Добавить элемент в стек
    let push value stack =
        Stack (value, stack)
    
    /// Прочитать головной элемент стека
    let top stack =
        match stack with
        | Empty -> None
        | Stack (element, _) -> Some(element)

    /// Удалить головной элемент стека
    let pop stack =
        match stack with
        | Empty -> None
        | Stack (_, residue) -> Some(residue)
    
    /// Проверка стека на пустоту 
    let isEmpty stack =
        (stack = Empty)

File: term_1 (C++, Algos and Data Structs)/homework_6/6.3/stack.cpp
#include "stack.h"
#include <iostream>

using std::string;

struct Element
{
	FlagType value;
	Element *next;
};

struct Stack
{
	Element *head;
};

Stack *createStack()
{
	return new Stack{ nullptr };
}

void push(Stack *&stack, FlagType value)
{
	Element *newElement = new Element{ value, stack->head };
	stack->head = newElement;
}

bool isEmpty(Stack *&stack)
{
	return stack->head == nullptr;
}

FlagType peek(Stack *&stack)
{
	return stack->head->value;
}

void pop(Stack *&stack)
{
	Element *position = stack->head;
	stack->head = position->next;
	delete position;
}

void deleteStack(Stack *&stack)
{
	while (!isEmpty(stack))
	{
		pop(stack);
	}
	delete stack;
	stack = nullptr;
}

string stackRemainder(Stack *&stack)
{
	string result = "";
	Element *position = stack->head;
	while (position)
	{
		result += position->value;
		position = position->next;
	}
	return result;
}